### PR TITLE
Remove 14MB from RunPeerCommit allocations

### PR DESCRIPTION
**Pull request by benaadams**
Rebased #5271 as was having issues with github 🤔

## Changes

- Only full refresh and parse from DB when nodes change
- Don't allocate fresh node prior to checking if node already exists (compare prior rather then post)
- Reduces by about ~10MB/min (see trace below)

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [ ] Yes
- [x] No

image

### CODE REVIEW

**Comment by benaadams** [comment]
Not sure why this change would crash the Test host process; isn't doing anything wild 🤔

File: src/Nethermind/Nethermind.Network/NetworkStorage.cs [hunk 1]
```diff
@@ -16,52 +20,93 @@ namespace Nethermind.Network
...
         public NetworkStorage(IFullDb? fullDb, ILogManager? logManager)
         {
             _fullDb = fullDb ?? throw new ArgumentNullException(nameof(fullDb));
             _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
         }
 
-        public int PersistedNodesCount
+        public int PersistedNodesCount => GetPersistedNodes().Length;
+
+        public NetworkNode[] GetPersistedNodes()
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/NetworkStorage.cs:40]: Does it have to return an Array? Can it just return IReadOnlyCollection and we could leverage the list itself? We could avoid storing array and list and allocating array all the time.

File: src/Nethermind/Nethermind.Network/NetworkStorage.cs [hunk 2]
```diff
@@ -16,52 +20,93 @@ namespace Nethermind.Network
...
+                    if (nodeRlp is null)
+                    {
+                        continue;
+                    }
+
+                    try
+                    {
+                        NetworkNode node = GetNode(nodeRlp);
+                        _nodesList.Add(node);
+                        _nodePublicKeys.Add(node.NodeId);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/NetworkStorage.cs:70]: We could potentially reuse already existing nodes instead of creating new ones all the time

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.Network/NetworkStorage.cs (hunk 1 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Network/NetworkStorage.cs:40]: Recreating the array gets around the concurrency problem of the list being used while the list is being adjusted since its used in a few places and the nodes are created via events from the network; so it would need tighter synchronisation on the iteration in `DiscoveryApp.AddPersistedNodes`; `NodesLoader.LoadPeersFromDb` and `PeerPool.UpdateReputationAndMaxPeersCount()` with the node creation form `ProtocolsManager.InitSyncPeerProtocol`; `DiscoveryApp.RunDiscoveryCommit()`; `DiscoveryManager.GetNodeLifecycleManager` and removal from `DiscoveryManager.OnIncomingMsg` and `PeerPool.CleanupPersistedPeers`

It doesn't have the issue currently as it completely recreates the `NetworkNode[]` from scratch from the DB each time the method is called; whereas here its reused form most method calls (and it only reads from the DB once for app lifetime, though it continues to write changes to it, but updates in memory structure instead of reading back)

**tl;dr** yes it could, but is complicated due to the concurrency

**Comment by benaadams** [comment]
Rebased

File: src/Nethermind/Nethermind.Network/NetworkStorage.cs (hunk 2 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Network/NetworkStorage.cs:70]: `RemoveNode` and `UpdateNode` don't set the `_nodes` array to `null` but just replace it so `GenerateNodes()` should only be creating the nodes from scratch once during app lifetime (i.e. this bit of code) and `UpdateNode` gets passed in the preconstructed `NetworkNode` so it would require a fair bit of refactoring to reuse the nodes from `RemoveNode` 🤔

**Comment by benaadams** [comment]
Not sure why this test failed
```
  Failed Should_count_total_coinbase_payments [435 ms]
  Error Message:
   Expected object to be equal to 100000000000, but found 200000000000.
```
MevRpc doesn't involve the peer nodes? 🤔

**Comment by benaadams** [comment]
Rebased again and squashed commits to see if persistent

File: src/Nethermind/Nethermind.Network/NetworkStorage.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Network/NetworkStorage.cs:70]: So rlp starts with public key, which could be decoded on stack, then it could be reused from existing node instead of creating new one.

We can say it is out of scope of this PR and try to do it in some next one.

File: src/Nethermind/Nethermind.Network/NetworkStorage.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Network/NetworkStorage.cs:40]: Fair enough, but you could have 2 lists that you would only copy from one to the other without any allocations or GC

File: src/Nethermind/Nethermind.Network/NetworkStorage.cs (hunk 2 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Network/NetworkStorage.cs:70]: Yeah I think would be good to look at that; as well as wider, as are a lot of temporary `byte[]` created for `Rlp` only to be fed into `KeccakHash.Compute` then thrown away; but the temp `byte[]` could be bypassed by feeding direct into `KeccakHash` which would also reduce copies by one

File: src/Nethermind/Nethermind.Network/NetworkStorage.cs (hunk 1 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Network/NetworkStorage.cs:40]: 🤔 could work; though use of the other list isn't guaranteed to have finished before starting to repopulate it

However thinking about that, could maybe get away with one List by taking a bounded Span reference via [`CollectionsMarshal.AsSpan<T>(List<T>)`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.asspan?view=net-7.0); then adding some null checks when the items are returned and skipping them. Would need to look into it a bit more as might be a bit wild 😅

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Network/NetworkStorage.cs:40]: We have this two list approach in `TxBroadcaster` with `_accumulatedTemporaryTxs` and `_txsToSend`

Comment (benaadams) [reply src/Nethermind/Nethermind.Network/NetworkStorage.cs:40]: Looks like its a single threaded reader and when the method is re-entered you know the iteration is complete? (e.g. timer fire); whereas here its read and written from a bunch of sources?

Comment (benaadams) [reply src/Nethermind/Nethermind.Network/NetworkStorage.cs:40]: Tried the 2 approach + Span, but is tricky to workout when to switch without it immediately switching back (e.g. `UpdateNodes(IEnumerable<NetworkNode> nodes)` calls `UpdateNode(NetworkNode node)` multiple times); and the Span still suffers from nodes being removed in the middle of the list, so skipping nodes 🤔

Comment (benaadams) [reply src/Nethermind/Nethermind.Network/NetworkStorage.cs:40]: Probably better to look at in future PR?

**Review by LukaszRozmej — APPROVED** [review]
